text/3502-cargo-script.md

```diff
@@ -0,0 +1,1202 @@
+[Configuring a target]: https://doc.rust-lang.org/cargo/guide/../reference/cargo-targets.html#configuring-a-target
+[def-package]:           https://doc.rust-lang.org/cargo/guide/../appendix/glossary.html#package          '"package" (glossary entry)'
+[Target auto-discovery]: https://doc.rust-lang.org/cargo/guide/../reference/cargo-targets.html#target-auto-discovery
+[TOML]: https://toml.io/
+[crates.io]: https://crates.io/
+[SemVer]: https://semver.org
+[def-crate]:             https://doc.rust-lang.org/cargo/guide/../appendix/glossary.html#crate             '"crate" (glossary entry)'
+[def-package]:           https://doc.rust-lang.org/cargo/guide/../appendix/glossary.html#package           '"package" (glossary entry)'
+[def-package-registry]:  https://doc.rust-lang.org/cargo/guide/../appendix/glossary.html#package-registry  '"package-registry" (glossary entry)'
+[def-manifest]:          https://doc.rust-lang.org/cargo/guide/../appendix/glossary.html#manifest          '"manifest" (glossary entry)'
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Single-file packages
+
+In addition to today's multi-file packages (`Cargo.toml` file with other `.rs`
+files),
+we are adding the concept of single-file packages which may contain an
+embedded manifest.
+There is no required way to distinguish a single-file
+`.rs` package from any other `.rs` file.
+
+A single-file package may contain an embedded manifest.
+An embedded manifest is stored using `TOML` in rust "frontmatter",
+a markdown code-fence with `cargo` at the start of the infostring at the top of
+the file.
+
+Inferred / defaulted manifest fields:
+- `package.name = <slugified file stem>`
+- `package.version = "0.0.0"` to [call attention to this crate being used in unexpected places](https://matklad.github.io/2021/08/22/large-rust-workspaces.html#Smaller-Tips)
+  - This may be altered with [rust-lang/cargo#9829](https://github.com/rust-lang/cargo/issues/9829)
+- `package.publish = false` to avoid accidental publishes, particularly if we
+  later add support for including them in a workspace.
+- `package.edition = <current>` to avoid always having to add an embedded
+  manifest at the cost of potentially breaking scripts on rust upgrades
+  - Warn when `edition` is unspecified.  Since piped commands are run with `--quiet`, this may not show up.
+  - Based on feedback, we might add `cargo-<edition>-edition` proxies to put in `#!` as a shorthand
+  - Based on feedback, we can switch to "edition is required as of <future> edition"
+
+Disallowed manifest fields:
+- `[workspace]`, `[lib]`, `[[bin]]`, `[[example]]`, `[[test]]`, `[[bench]]`
+- `package.workspace`, `package.build`, `package.links`, `package.publish`, `package.autobins`, `package.autoexamples`, `package.autotests`, `package.autobenches`
+
+Single-file packages maintain an out-of-line target directory by default.
+This is `$CARGO_HOME/target/<hash-of-path>`.
```

Comment:
This means we lose the target dir when we rename or move the script. As `<hash-of-path>` is unreadable anyway how do we feel about using something else?

A files inode number (POSIX) or File ID (Windows) stays the same between renames and moves on the same filesystem. Combining inode and a filesystem description results in a system unique identifier.

This could look like `$CARGO_HOME/target/dev/sda/232389`. 

Further more to clean up target directories for deleted *cargo-scripts* we can check if a *cargo-script* still exists at that inode (the numbers are reused as files are deleted).

Reply:
It's very common for editors to replace a file with a new one on each change, to get atomic updates, which would mean that every time you edit the script it gets a new target directory and has to rebuild all dependencies.

Reply:
I didnt know that, thanks for bringing it up! That rules out innodes.

I would still like to be able to move a *cargo-script* around without having to recompile. Its not *that* important to me though.

The only other approach I can think of right now is to take as little of the path as possible and expand on hash collisions. I do not think the extra complexity that would lead to is worth it though...

Reply:
While I've not dived into cargo fingerprinting all that much, from what I understand, moving a package around will already cause a lot ot be recompiled.  I see #5931 as the way to be reducing recompiles.

Reply:
I called this out in 93e8b1f